lecture_3/03_project/calculator/internal/calcengine/calc_engine.go

```diff
@@ -0,0 +1,61 @@
+package calcengine
+
+import (
+	"context"
+	"log"
+)
+
+
+type Engine struct {
+	consumer  Consumer
+	handler   Handler
+	publisher Publisher
+}
+
+func New(consumer Consumer, handler Handler, publisher Publisher) *Engine {
+	return &Engine{
+		consumer:  consumer,
+		handler:   handler,
+		publisher: publisher,
+	}
+}
+
+func (e *Engine) Start(ctx context.Context) {
+	err := e.processEventUpdates(ctx)
+	if err != nil {
+		log.Println("Engine failed to process event updates:", err)
+		return
+	}
+
+	err = e.processBets(ctx)
+	if err != nil {
+		log.Println("Engine failed to process bets:", err)
+		return
+	}
+
+	<-ctx.Done()
+}
+
+func (e *Engine) processEventUpdates(ctx context.Context) error {
+	consumedEventUpdates, err := e.consumer.ConsumeEventUpdates(ctx)
+	if err != nil {
+		return err
+	}
+
+	newBets := e.handler.HandleEventUpdates(ctx, consumedEventUpdates)
+	e.publisher.PublishBetsCalculated(ctx, newBets)
+
+	return nil
+}
+
+func (e *Engine) processBets(ctx context.Context) error {
+	consumedBets, err := e.consumer.ConsumeBets(ctx)
+	if err != nil {
+		return err
+	}
+
+	newBets := e.handler.HandleBets(ctx, consumedBets)
```

Comment:
This channel is always empty. I would suggest changing this method signature, so that it does not return a channel of bets because it is a little bit confusing.